### ISSUE TITLE
Decrease memory usage of SftpAdapter->listDirectoryContents().

### ISSUE THREAD

**Issue by StefanPrintezis** [thephpleague/flysystem#1077]
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| BC Break    | no

#### Scenario
Listing the contents of a directory with a 100k files results in very high memory usage, which is not needed when you want to process each file individually.

#### Summary
I've created a pull request over at phpseclib for adding a generator that 'streams' the files as they are retrieved by several sftp requests.
https://github.com/phpseclib/phpseclib/pull/1418
Eventually if that's accepted we'd need to implement that here.

What it could look like:

```
  /**
     * List the contents of a directory as a generator.
     *
     * @param string $directory
     * @param bool   $recursive // Not sure what to do with recursion.
     *
     * @return \Generator
     */
    protected function listDirectoryContentsGenerator($directory, $recursive = true)
    {
        $result = [];
        $connection = $this->getConnection();
        $location = $this->prefix($directory);
        $listing = $connection->rawlistGenerator($location);

        if ($listing === false) {
            return [];
        }

        foreach ($listing as $file) {
            // When directory entries have a numeric filename they are changed to int
            $filename = $file['filename'] ?? '';
            if (in_array($filename, ['.', '..'])) {
                continue;
            }

            $path = empty($directory) ? $filename : ($directory.'/'.$filename);

            yield $this->normalizeListingObject($path, $object);
            
            // if ($recursive && isset($object['type']) && $object['type'] === NET_SFTP_TYPE_DIRECTORY) {
               // $result = array_merge($result, $this->listDirectoryContents($path));
            // }
        }
    }

**Comment by garygreen** [comment]
This should be closed in favour of: #970

**Comment by StefanPrintezis** [comment]
The solution mentioned in #970 does not affect the sftp adapter that requires a change in the phpseclib i'm working on. We could combine them but since they're different adapters it could be a different issue.
